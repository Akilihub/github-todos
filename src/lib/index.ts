/**
 * create issues out of TODO and FIXME comments if they don't exist yet
 * with in a repo issue.
 */

import getChangedFiles from "../lib/github/changedFilesList";
import getFilesContent, { FilesContent } from "./github/getFilesContent";
import getNewIssuesFromSource from "./github/getNewIssue";
import addIssuesToRepo from "./github/addIssuesToRepo";
import getFiles from "../lib/github/getFilesOnInstall";
import parseData from "./parser";

import { mergeFileRepoIssues } from "./github/utils";

import { Context } from "./github/types";

export async function addRepoCommentsToGH(context: Context): Promise<void> {
    // get changed files on a git push
    const changedFiles = await getChangedFiles(context);
    // get file data
    const filesContent: FilesContent[] = await getFilesContent(changedFiles); // error handling

    const filesRepoIssues = filesContent.map(parseData);
    // merge into a single set of FIXME and TODO issues
    const repoIssues = mergeFileRepoIssues(filesRepoIssues);
    const issuesToAddToRepo = await getNewIssuesFromSource(context, repoIssues);
    return addIssuesToRepo(context, issuesToAddToRepo);
}

export async function addCommentsToGHonInstall(context: Context): Promise<void> {
    // get changed files on a git push
    const changedFiles = await getFiles(context);
    // get file data
    const filesContent: FilesContent[] = await getFilesContent(changedFiles); // error handling
    // FIXME: this doesnot work
    // Just ignore it
    const filesRepoIssues = filesContent.map(parseData);
    // merge into a single set of FIXME and TODO issues
    const repoIssues = mergeFileRepoIssues(filesRepoIssues);
    const issuesToAddToRepo = await getNewIssuesFromSource(context, repoIssues);
    return addIssuesToRepo(context, issuesToAddToRepo);
}

