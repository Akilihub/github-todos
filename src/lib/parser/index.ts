/**
 * uses an AST parser to analyse commits
 * currently supports ts and js files
 */
import extract from "esprima-extract-comments";

import { FilesContent } from "../github/getFilesContent";

type KeyWord = "TODO"  | "FIXME";

const keyWords: KeyWord[] = ["TODO", "FIXME"];

export interface RepoIssue {
   commentText: string;
   lineNumber: number;
   keyWord: KeyWord;
   fileName: string;
   url: string;
}

interface ExtractedBlob {
    type: string;
    value: string;
    range: number[];
    loc: any;
}

const findWord = (str: string): KeyWord | null => {
  const wordArr = keyWords.filter(word => str.includes(word));
  return wordArr ?  wordArr[0] : null;
};

const replaceWord = (str: string): string => {
    const regex = /[TODO:FIXME:]/g;
    return str.replace(regex, "").trim();
};

function parseData (data: FilesContent): RepoIssue[] {
    const extractedFile: ExtractedBlob[] = extract(data.content);
    return extractedFile
    .map( (obj: ExtractedBlob) => ({
        commentText: replaceWord(obj.value),
        lineNumber: obj.loc.end.line,
        keyWord: findWord(obj.value) as KeyWord,
        fileName: data.name,
        url: data.url
    }))
    .filter(obj => keyWords.includes(obj.keyWord));
}

export default parseData;
