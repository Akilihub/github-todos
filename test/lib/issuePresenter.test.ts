import "jest";
import formattedIssueComment from "../../src/lib/github/issuePresenter";
import { RepoIssue } from "../../src/lib/parser";
const data: RepoIssue[] = [
    {
      "commentText": "@aleku399 look into why this was causing bars not to show",
      "fileName": "README.md",
      "keyWord": "TODO",
      "lineNumber": 12,
      "url": "https://github.com/epicallan/repo/"
    },
    {
      "commentText": "alex testing 1 on the 2s",
      "fileName": "README.md",
      "keyWord": "FIXME",
      "lineNumber": 14,
      "url": "https://github.com/epicallan/repo/"
    },
    {
      "commentText": "test",
      "fileName": "README.md",
      "keyWord": "TODO",
      "lineNumber": 11,
      "url": "https://github.com/epicallan/repo/"
    }
  ];
describe("parsed data tests", () => {
    it("should return modified parsed data", () => {
      const result = formattedIssueComment(data);
      expect(result).toMatchSnapshot();
    });
});