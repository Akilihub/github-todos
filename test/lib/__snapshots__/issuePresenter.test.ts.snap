// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsed data tests should return modified parsed data 1`] = `
Array [
  Object {
    "body": "- [ ] [README.md](https://github.com/epicallan/repo/#L12) : @aleku399 look into why this was causing bars not to show
- [ ] [README.md](https://github.com/epicallan/repo/#L11) : test",
    "title": "TODOs from source code",
  },
  Object {
    "body": "- [ ] [README.md](https://github.com/epicallan/repo/#L14) : alex testing 1 on the 2s",
    "title": "FIXMEs from source code",
  },
]
`;
